clients/admin-ui/src/features/common/custom-fields/hooks.ts

```diff
@@ -0,0 +1,191 @@
+    useDeleteCustomFieldMutation({ fixedCacheKey: resourceFidesKey });
+
+  const isLoading =
+    allAllowListQuery.isLoading ||
+    customFieldDefinitionsQuery.isLoading ||
+    customFieldsQuery.isLoading ||
+    upsertCustomFieldMutationResult.isLoading ||
+    deleteCustomFieldMutationResult.isLoading;
+
+  const idToAllowListWithOptions = useMemo(
+    () =>
+      new Map(
+        filterWithId(allAllowListQuery.data).map((allowList) => [
+          allowList.id,
+          {
+            ...allowList,
+            options: (allowList.allowed_values ?? []).map((value) => ({
+              value,
+              label: value,
+            })),
+          },
+        ])
+      ),
+    [allAllowListQuery.data]
+  );
+
+  const idToCustomFieldDefinition = useMemo(
+    () =>
+      new Map(
+        filterWithId(customFieldDefinitionsQuery.data).map((cfd) => [
+          cfd.id,
+          cfd,
+        ])
+      ),
+    [customFieldDefinitionsQuery]
+  );
+
+  const definitionIdToCustomField = useMemo(
+    () =>
+      new Map(
+        filterWithId(customFieldsQuery.data).map((fd) => [
+          fd.custom_field_definition_id,
+          fd,
+        ])
+      ),
+    [customFieldsQuery]
+  );
+
+  const sortedCustomFieldDefinitionIds = useMemo(() => {
+    const ids = [...idToCustomFieldDefinition.keys()];
+    ids.sort();
+    return ids;
+  }, [idToCustomFieldDefinition]);
+
+  /**
+   * Issue a batch of upsert and delete requests that will sync the form selections to the
+   * backend.
+   */
+  const upsertCustomFields = useCallback(
+    async (formValues: CustomFieldsFormValues) => {
+      // When creating an resource, the fides key may have initially been blank. But by the time the
+      // form is submitted it must not be blank (not undefined, not an empty string).
+      const fidesKey = formValues.fides_key || resourceFidesKey;
+      if (!fidesKey) {
+        return;
+      }
+
+      const { definitionIdToCustomFieldValue } = formValues;
+
+      // This will be undefined if the form never rendered a `CustomFieldList` that would assign
+      // form values.
+      if (!definitionIdToCustomFieldValue) {
+        return;
+      }
+
+      try {
+        // This would be a lot simpler (and more efficient) if the API had an endpoint for updating
+        // all the metadata associated with a field, including deleting options that weren't passed.
+        await Promise.allSettled(
+          sortedCustomFieldDefinitionIds.map((definitionId) => {
+            const customField = definitionIdToCustomField.get(definitionId);
+            const value = definitionIdToCustomFieldValue[definitionId];
+
+            if (
+              value === undefined ||
+              value === "" ||
+              (Array.isArray(value) && value.length === 0)
+            ) {
+              if (!customField?.id) {
+                return undefined;
+              }
+
+              return deleteCustomFieldMutationTrigger(customField);
```

Comment:
The previous submission code didn't have the ability to delete custom fields. I think the API forbids setting the field to an empty string, which makes sense because that isn't an allowed value. So we have to delete the field assignment if the value is blank.

Reply:
The API does allow for an empty array to be submitted. However, I do understand about your point about deleting the custom field assignment.